torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp

```diff
@@ -567,256 +370,390 @@ Value* findArgumentAsInputParam(
   }
-  fullName += name;
-
-  if (allAttrValues.find(fullName) == allAttrValues.end() &&
-      attrModule.hasattr(name)) {
-    auto attr = attrModule.attr(name);
-    auto type = attrModule.type();
-    auto slot = *type->findAttributeSlot(name);
-
-    // Add model_parameters and model_buffers as model inputs. Order is
-    // preserved based on the appearance in the graph.
-    if (type->is_parameter(slot) || type->is_buffer(slot) ||
-        (attr.isObject() && !attr.toObjectRef().type()->is_module())) {
-      if (allAttrValues.find(fullName) == allAttrValues.end()) {
-        paramConst = findArgumentAsInputParam(graph, fullName, attr);
-        allAttrValues.insert({fullName, paramConst});
-      }
-    } else if (auto attrVal = tryInsertConstant(*graph, attr)) {
-      for (size_t i = 0; i < type->getAttributes().size(); i++) {
-        if (type->getAttributeName(i) == name) {
-          paramConst = *attrVal;
-          allAttrValues.insert({fullName, paramConst});
+
+  auto root_v = alias_to_value_[old_v];
+  alias_to_value_[new_v] = root_v;
+  auto& sorted_alias = value_to_sorted_aliases_[root_v];
+  sorted_alias.insert(new_v);
+
+  // check if new_v is created inside if or loop subblock.
+  auto* owning_blocknode = owning_block->owningNode();
+  if (nullptr == owning_blocknode) {
+    return;
+  }
+  auto owning_block_nkind = owning_blocknode->kind();
+  if (owning_block_nkind != prim::Loop && owning_block_nkind != prim::If) {
+    return;
+  }
+
+  bool registered = std::any_of(
+      owning_block->outputs().begin(),
+      owning_block->outputs().end(),
+      [&sorted_alias](Value* out) {
+        return std::any_of(
+            sorted_alias.begin(), sorted_alias.end(), [&out](Value* alias) {
+              return alias == out;
+            });
+      });
+
+  bool from_outer_alias = std::any_of(
+      sorted_alias.begin(),
+      sorted_alias.end(),
+      [&owning_blocknode](Value* alias) {
+        return isAncestor(
+            alias->node()->owningBlock(), owning_blocknode->owningBlock());
+      });
+
+  // The data of this value has been changed.
+  // If this value has alias from outer block,
+  // then the update must be reflected back to outside.
+  // Thus it needs to be registered as a subblock output.
+  // This step can be skipped if other alias of this value has already been registered as sublock output.
+  if (!registered && from_outer_alias) {
+    if (owning_block_nkind == prim::Loop) {
+      owning_block->registerOutput(new_v);
+      auto new_block_in = owning_block->addInput();
+      sorted_alias.insert(new_block_in);
+      alias_to_value_[new_block_in] = root_v;
+      owning_blocknode->addInput(root_v);
+    } else if (owning_block_nkind == prim::If) {
+      for (auto* if_sub_block : owning_blocknode->blocks()) {
+        if (owning_block == if_sub_block) {
+          if_sub_block->registerOutput(new_v);
+        } else {
+          if_sub_block->registerOutput(root_v);
```

Comment:
Can we register an output from outside of the block? Do we need a clone?

Reply:
Yes we can, clone(Identity) is added in `fixup_onnx_controlflow` pass